### PR TITLE
Parameterize google rtc audio processing

### PR DESCRIPTION
**Pull request by perahgren**
Changes in the GoogleRtcAudioProcessing component to
-Allow sample rate and number of channels to explicitly be specified.  
-Update the memory management.
-Allow parameters to be passed.

### CODE REVIEW

File: src/audio/google_rtc_audio_processing.c [hunk 1]
```diff
@@ -225,14 +330,36 @@ static struct comp_dev *google_rtc_audio_processing_create(
 	if (!cd->tuning_handler)
 		goto fail;
 
-	cd->state = GoogleRtcAudioProcessingCreate();
+	if (MEMORY_BUFFER_BYTES > 0) {
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing.c:333]: always true as this is hard coded

File: src/audio/google_rtc_audio_processing.c [hunk 2]
```diff
@@ -225,14 +330,36 @@ static struct comp_dev *google_rtc_audio_processing_create(
...
+					    MEMORY_BUFFER_BYTES * sizeof(cd->memory_buffer[0]));
+		if (!cd->memory_buffer) {
+			comp_err(dev, "google_rtc_audio_processing_create: failed to allocate memory buffer");
+
+			goto fail;
+		}
+
+		GoogleRtcAudioProcessingAttachMemoryBuffer(cd->memory_buffer, MEMORY_BUFFER_BYTES);
+	} else {
+		cd->memory_buffer = NULL;
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing.c:344]: this should be impossible to get to, and if you did modify ti to be accessible you will leak here

File: src/audio/google_rtc_audio_processing.c [hunk 3]
```diff
@@ -275,6 +402,10 @@ static struct comp_dev *google_rtc_audio_processing_create(
 	if (cd) {
 		rfree(cd->output_buffer);
 		rfree(cd->aec_reference_buffer);
+		if (cd->memory_buffer) {
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing.c:405]: not sure about your API call but memory buffer does not need to be in the if as rfree(NULL) is a no-op

File: src/audio/google_rtc_audio_processing_mock.c [hunk 4]
```diff
@@ -45,6 +54,35 @@ GoogleRtcAudioProcessingState *GoogleRtcAudioProcessingCreate()
...
+									int config_size)
+{
+	struct GoogleRtcAudioProcessingState *s =
+		rballoc(0, SOF_MEM_CAPS_RAM, sizeof(GoogleRtcAudioProcessingState));
+	if (!s)
+		return NULL;
+	s->num_capture_channels = num_capture_input_channels;
+	s->num_aec_reference_channels = num_render_channels;
+	s->num_output_channels = num_capture_output_channels;
+	s->num_frames = capture_sample_rate_hz * 10 / 1000;
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing_mock.c:72]: what is 10/1000? is it a time conversion?

File: src/audio/google_rtc_audio_processing_mock.c [hunk 5]
```diff
@@ -53,6 +91,35 @@ void GoogleRtcAudioProcessingFree(GoogleRtcAudioProcessingState *state)
...
+	s->aec_reference = rballoc(0,
+				   SOF_MEM_CAPS_RAM,
+				   sizeof(s->aec_reference[0]) *
+				   s->num_frames *
+				   s->num_aec_reference_channels
+				   );
+	return 0;
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing_mock.c:113]: this looks very similar to above, can we maybe reuse it with helpers or inlines to mask out variables?

File: src/audio/google_rtc_audio_processing_mock.c [hunk 6]
```diff
@@ -94,3 +161,30 @@ int GoogleRtcAudioProcessingAnalyzeRender_int16(
 		 data, buffer_size);
 	return 0;
 }
+
+void GoogleRtcAudioProcessingParseSofConfigMessage(uint8_t *message,
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing_mock.c:165]: this does not appear to be parsing

Comment (perahgren) [reply src/audio/google_rtc_audio_processing_mock.c:171]: No, the mock does not do the parsing and instead just flags that no parameters were specified.

File: src/audio/google_rtc_audio_processing_mock.c (hunk 4 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing_mock.c]: 10 is the frame size in ms, and the computation produces in the frame size in samples.
This now changed this to make it more clear.

File: src/audio/google_rtc_audio_processing.c (hunk 2 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c:344]: I don't follow.
If MEMORY_BUFFER is zero, this simply initializes memory_buffer to NULL, and no allocation is done of memory_buffer. Therefore there should not be any leakage.

File: src/audio/google_rtc_audio_processing.c [hunk 7]
```diff
@@ -34,8 +34,18 @@
...
-#define GOOGLE_RTC_AUDIO_PROCESSING_SAMPLERATE 48000
+#define SAMPLE_RATE_HZ 48000
+#define SAMPLES_PER_CHANNEL (SAMPLE_RATE_HZ / 100)
+#define DEFAULT_NUM_AEC_REFERENCE_CHANNELS 2
+#define DEFAULT_NUM_CAPTURE_INPUT_CHANNELS 1
+#define DEFAULT_NUM_CAPTURE_OUTPUT_CHANNELS 1
+
+#define MEMORY_BUFFER_STEREO_SIZE_BYTES 250000
+#define MEMORY_BUFFER_MONO_SIZE_BYTES 200000
+#define MEMORY_BUFFER_BYTES (DEFAULT_NUM_CAPTURE_OUTPUT_CHANNELS == 1 ? \
```
[the remark this answers is not in the document]
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c:47]: Yes. But my concern with that is that this size needs to be defined compile-time, so if a macro is defined, it mustn't be used on any runtime-specified parameters. 
With that I'm not convinced that adding a macro is the correct thing to do.

**Comment by marc-hb** [comment]
@perahgren since you're doing "rapide-fire force pushes" I change the PR title and added the (magic and cryptic, sorry) [SKIP CI] prefix to stop the saturation of our limited hardware resources at sof-ci/jenkins/pr-device-test

File: src/audio/google_rtc_audio_processing.c [hunk 8]
```diff
@@ -34,8 +34,18 @@
 
 #include <google_rtc_audio_processing.h>
 #include <google_rtc_audio_processing_platform.h>
+#include <google_rtc_audio_processing_sof_message_reader.h>
 
-#define GOOGLE_RTC_AUDIO_PROCESSING_SAMPLERATE 48000
+#define SAMPLE_RATE_HZ 48000
+#define SAMPLES_PER_CHANNEL (SAMPLE_RATE_HZ / 100)
```
Comment (lgirdwood) [inline src/audio/google_rtc_audio_processing.c:40]: I would be careful with these generic names, best to use the prefix to prevent collisions with global names.

Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c]: Thanks for the headsup! I've now added the GOOGLE_RTC_AUDIO_PROCESSING prefix to avoid that

File: src/audio/google_rtc_audio_processing.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (cujomalainey) [reply src/audio/google_rtc_audio_processing.c]: You could add a complie time assert on the final calculated value to make sure it is compile time available. Also why does it need to be compile time ready?

File: src/audio/google_rtc_audio_processing.c (hunk 1 above)
Comment (cujomalainey) [reply src/audio/google_rtc_audio_processing.c]: It's written as if it can chance at runtime but that is weird given it's a constant

File: src/audio/google_rtc_audio_processing_mock.c (hunk 6 above)
Comment (cujomalainey) [reply src/audio/google_rtc_audio_processing_mock.c:172]: Haha missed I moved into the mock file, thanks for the pointer

File: src/audio/google_rtc_audio_processing.c [hunk 9]
```diff
@@ -34,8 +34,19 @@
 
 #include <google_rtc_audio_processing.h>
 #include <google_rtc_audio_processing_platform.h>
+#include <google_rtc_audio_processing_sof_message_reader.h>
 
-#define GOOGLE_RTC_AUDIO_PROCESSING_SAMPLERATE 48000
+#define GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ 48000
+#define GOOGLE_RTC_AUDIO_PROCESSING_SAMPLES_PER_CHANNEL (GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ / 100)
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing.c:40]: PERIOD_SAMPLES_PER_CHANNEL

File: src/audio/google_rtc_audio_processing.c [hunk 10]
```diff
@@ -34,8 +34,19 @@
...
+#define GOOGLE_RTC_AUDIO_PROCESSING_DEFAULT_NUM_AEC_REFERENCE_CHANNELS 2
+#define GOOGLE_RTC_AUDIO_PROCESSING_DEFAULT_NUM_CAPTURE_INPUT_CHANNELS 1
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing.c:42]: How hard will it be to make this dynamic?

File: src/audio/google_rtc_audio_processing.c [hunk 11]
```diff
@@ -113,11 +127,104 @@ static int google_rtc_audio_processing_reconfigure(struct comp_dev *dev)
...
+						      &num_capture_output_channels,
+						      &aec_reference_delay,
+						      &mic_gain,
+						      &google_rtc_audio_processing_config_present,
+						      &num_capture_input_channels_present,
+						      &num_capture_output_channels_present,
+						      &aec_reference_delay_present,
+						      &mic_gain_present);
+
+	if (num_capture_input_channels_present || num_capture_output_channels_present) {
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing.c:156]: isnt the config invalid if these are 0?

File: src/audio/google_rtc_audio_processing.c [hunk 12]
```diff
@@ -113,11 +127,104 @@ static int google_rtc_audio_processing_reconfigure(struct comp_dev *dev)
...
+							       GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ,
+							       cd->num_capture_input_channels,
+							       cd->num_capture_output_channels,
+							       GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ,
+							       cd->num_aec_reference_channels);
+
+		if (ret) {
+			comp_err(dev, "GoogleRtcAudioProcessingSetStreamFormats failed: %d",
+				 ret);
+			return -EINVAL;
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing.c:180]: is the library returning a linux errno? if yes we should just return that

File: src/audio/google_rtc_audio_processing.c [hunk 13]
```diff
@@ -113,11 +127,104 @@ static int google_rtc_audio_processing_reconfigure(struct comp_dev *dev)
...
+
+			comp_info(dev,
+				  "google_rtc_audio_processing_reconfigure(): Applying capture linear headroom: %d.%d",
+				  (int)mic_gain, (int)(100 * mic_gain) - 100 * ((int)mic_gain));
+		}
+		if (aec_reference_delay_present) {
+			echo_path_delay_ms_use = &aec_reference_delay;
+			comp_info(dev,
+				  "google_rtc_audio_processing_reconfigure(): Applying aec reference delay: %d.%d",
+				  (int)aec_reference_delay,
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing.c:199]: not sure i grasp what is going on here, you appear to be logging a pointer, is that intended?

File: src/audio/google_rtc_audio_processing.c [hunk 14]
```diff
@@ -325,7 +460,7 @@ static int google_rtc_audio_processing_prepare(struct comp_dev *dev)
 
 		if (source_c->source->pipeline->pipeline_id != dev->pipeline->pipeline_id) {
 			cd->aec_reference = source;
-			aec_channels = sourcs_c->stream.channels;
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing.c:328]: this should be its own commit so it can be cherrypicked without your change if needed

File: src/audio/google_rtc_audio_processing_mock.c [hunk 15]
```diff
@@ -22,29 +23,79 @@ struct GoogleRtcAudioProcessingState {
...
+		       int num_capture_input_channels,
+		       int num_capture_output_channels,
+		       int render_sample_rate_hz,
+		       int num_render_channels)
+{
+	s->num_capture_channels = num_capture_input_channels;
+	s->num_output_channels = num_capture_output_channels;
+	s->num_frames = capture_sample_rate_hz / GOOGLE_RTC_AUDIO_PROCESSING_FRAMES_PER_SECOND;
+
+	if (s->num_aec_reference_channels != num_render_channels ||
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing_mock.c:37]: brackets for clarity on logic

**Review by cujomalainey — CHANGES_REQUESTED** [review]

File: src/audio/google_rtc_audio_processing.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c]: Yes, I think you are right, and my reasons for keeping this as a #define does not really hold.
I changed it to a function instead though (because I would prefer to not use a macro).

The information needs to be available at the call to _create though. However, if the buffer allocated is too small, the functionality fallbacks to using the SOF allocation routines.

File: src/audio/google_rtc_audio_processing.c (hunk 1 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c]: The code is designed to work also when setting either GOOGLE_RTC_AUDIO_PROCESSING_MEMORY_BUFFER_STEREO_SIZE_BYTES
GOOGLE_RTC_AUDIO_PROCESSING_MEMORY_BUFFER_MONO_SIZE_BYTES to zero.

It looks slightly different now though

File: src/audio/google_rtc_audio_processing.c (hunk 2 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c]: Nice! I removed that.

File: src/audio/google_rtc_audio_processing.c (hunk 9 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c]: Done

File: src/audio/google_rtc_audio_processing.c (hunk 10 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c:42]: It is doable, but there are some requirements involved.

The trickiest part is that these needs to be available at _create, which means that when they are needed, there is yet no way to pass values from outside of the firmware.

File: src/audio/google_rtc_audio_processing.c (hunk 11 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c:161]: No, these values are only optional in the config. 
The config can now carry different optional payloads, and the presence of those is flagged using the bools
The channel counts passed in the config can be 0, but such will be rejected by GoogleRtcAudioProcessingSetStreamFormats (and an error code will be returned)

File: src/audio/google_rtc_audio_processing.c (hunk 12 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c]: Good point! Done.

File: src/audio/google_rtc_audio_processing.c (hunk 13 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c:204]: What I wanted to achieve is to print out a floating point value.
However, it seems like with comp_info, using #f for floating point values is broken, and just printing random values.

This was the best I could do to achieve a log-line with a floating point value

File: src/audio/google_rtc_audio_processing.c (hunk 14 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c]: Done. Reverted my change.

File: src/audio/google_rtc_audio_processing_mock.c (hunk 15 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing_mock.c]: Like this?

File: src/audio/google_rtc_audio_processing.c (hunk 10 above)
Comment (cujomalainey) [reply src/audio/google_rtc_audio_processing.c:42]: Got it, in the meantime how about kconfig? We should probably also have checks to make sure the pipeline isn't outside out spec

File: src/audio/google_rtc_audio_processing.c (hunk 13 above)
Comment (cujomalainey) [reply src/audio/google_rtc_audio_processing.c:204]: Might be worth adding a comment since its not particularly clear

File: src/audio/google_rtc_audio_processing.c [hunk 16]
```diff
@@ -225,14 +337,36 @@ static struct comp_dev *google_rtc_audio_processing_create(
...
 		goto fail;
 
-	cd->state = GoogleRtcAudioProcessingCreate();
+	cd->num_aec_reference_channels = GOOGLE_RTC_AUDIO_PROCESSING_DEFAULT_NUM_AEC_REFERENCE_CHANNELS;
+	cd->num_capture_input_channels = GOOGLE_RTC_AUDIO_PROCESSING_DEFAULT_NUM_CAPTURE_INPUT_CHANNELS;
+	cd->num_capture_output_channels = GOOGLE_RTC_AUDIO_PROCESSING_DEFAULT_NUM_CAPTURE_OUTPUT_CHANNELS;
+	cd->num_frames = GOOGLE_RTC_AUDIO_PROCESSING_PERIOD_SAMPLES_PER_CHANNEL;
+
+	int memory_buffer_size_bytes = get_memory_buffer_size_bytes(cd->num_capture_output_channels);
+	if (memory_buffer_size_bytes > 0) {
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing.c:346]: based on the function call, this appears to be always true

File: src/audio/google_rtc_audio_processing.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (cujomalainey) [reply src/audio/google_rtc_audio_processing.c]: My biggest issue is that it detaches the pipeline expectations from the component. I think it would be wiser to have the component check the pipeline against its needs rather than assume based on a compilation config

[the remark this answers is not in the document]
Comment (cujomalainey) [reply src/audio/google_rtc_audio_processing.c]: That being said, we do have the issue right now where DMIC is more channels than we actually care about. Might be worth fixing up a demux properly. @sathya-nujella FYI

File: src/audio/google_rtc_audio_processing_mock.c (hunk 4 above)
Comment (cujomalainey) [reply src/audio/google_rtc_audio_processing_mock.c]: Might be worth defineing that as it is not inherently clear

File: src/audio/google_rtc_audio_processing_mock.c [hunk 17]
```diff
@@ -22,29 +23,79 @@ struct GoogleRtcAudioProcessingState {
...
+		       int num_render_channels)
+{
+	s->num_capture_channels = num_capture_input_channels;
+	s->num_output_channels = num_capture_output_channels;
+	s->num_frames = capture_sample_rate_hz / GOOGLE_RTC_AUDIO_PROCESSING_FRAMES_PER_SECOND;
+
+	if ((s->num_aec_reference_channels != num_render_channels) ||
+	    !s->aec_reference) {
+		s->num_aec_reference_channels = num_render_channels;
+		if (s->aec_reference) {
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing_mock.c:40]: if not needed

Comment (perahgren) [reply src/audio/google_rtc_audio_processing_mock.c:40]: True. Removed that.

File: src/audio/google_rtc_audio_processing.c (hunk 13 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c:204]: Good point.
Done

File: src/audio/google_rtc_audio_processing_mock.c (hunk 4 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing_mock.c]: Done.

File: src/audio/google_rtc_audio_processing.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c]: The scheme here is fairly easily extended to react on the pipeline needs.
Right now, however, the component _copy code is so hardcoded to always process mono mic channels, so I'd prefer to postpone that change until the _copy code is properly extended.

File: src/audio/google_rtc_audio_processing.c (hunk 16 above)
Comment (cujomalainey) [reply src/audio/google_rtc_audio_processing.c]: I think 1 is fine but then it needs to be clear it can be modified, e.g. an optional kconfig

**Review by cujomalainey — APPROVED** [review]
one nit, also approval with the agreement we move away from hard coded defines in the future

File: src/audio/google_rtc_audio_processing_mock.c [hunk 18]
```diff
@@ -22,29 +24,76 @@ struct GoogleRtcAudioProcessingState {
...
-	s->num_capture_channels = 1;
-	s->num_aec_reference_channels = 2;
-	s->num_output_channels = 1;
-	s->num_frames = GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ * 10 / 1000;
+	s->num_capture_channels = num_capture_input_channels;
+	s->num_output_channels = num_capture_output_channels;
+	s->num_frames = capture_sample_rate_hz / GOOGLE_RTC_AUDIO_PROCESSING_FRAMES_PER_SECOND;
+
+	s->num_aec_reference_channels = num_render_channels;
+	if (s->aec_reference) {
```
[the remark this answers is not in the document]
Comment (perahgren) [reply src/audio/google_rtc_audio_processing_mock.c:39]: I think that is needed, since SetFormats may be called with it being NULL: i.e., from line 72.

[the remark this answers is not in the document]
Comment (perahgren) [reply src/audio/google_rtc_audio_processing_mock.c]: Well spotted. That indeed was wrong.
Done

File: src/audio/google_rtc_audio_processing_mock.c [hunk 19]
```diff
@@ -22,29 +24,76 @@ struct GoogleRtcAudioProcessingState {
...
+		       int render_sample_rate_hz,
+		       int num_render_channels)
+{
+	state->num_capture_channels = num_capture_input_channels;
+	state->num_output_channels = num_capture_output_channels;
+	state->num_frames = capture_sample_rate_hz / GOOGLE_RTC_AUDIO_PROCESSING_FRAMES_PER_SECOND;
+
+	state->num_aec_reference_channels = num_render_channels;
+	if (state->aec_reference) {
+		rfree(state->aec_reference);
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing_mock.c:40]: `if` isn't needed

Comment (perahgren) [reply src/audio/google_rtc_audio_processing_mock.c:40]: I think it is needed, since SetFormats could both be called with state->aec_reference allocated, and being NULL.

E.g., on line 72 it is NULL, and on line 112 it is allocated.

File: src/audio/google_rtc_audio_processing_mock.c [hunk 20]
```diff
@@ -22,29 +24,74 @@ struct GoogleRtcAudioProcessingState {
...
+	state->num_output_channels = num_capture_output_channels;
+	state->num_frames = capture_sample_rate_hz / GOOGLE_RTC_AUDIO_PROCESSING_FRAMES_PER_SECOND;
+
+	state->num_aec_reference_channels = num_render_channels;
+	rfree(state->aec_reference);
+	state->aec_reference = rballoc(0,
+				       SOF_MEM_CAPS_RAM,
+				       sizeof(state->aec_reference[0]) *
+				       state->num_frames *
+				       state->num_aec_reference_channels);
```
Comment (lyakh) [inline src/audio/google_rtc_audio_processing_mock.c:44]: would `rbrealloc()` work for you?

File: src/audio/google_rtc_audio_processing_mock.c [hunk 21]
```diff
@@ -53,21 +100,46 @@ void GoogleRtcAudioProcessingFree(GoogleRtcAudioProcessingState *state)
...
+	return state->num_frames *
+		GOOGLE_RTC_AUDIO_PROCESSING_MS_PER_SECOND /
+		GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ;
 }
 
 int GoogleRtcAudioProcessingReconfigure(
-	GoogleRtcAudioProcessingState * const state, const uint8_t *const config,
-	int config_size)
+					GoogleRtcAudioProcessingState * const state, const uint8_t *const config,
+					int config_size)
```
Comment (lyakh) [inline src/audio/google_rtc_audio_processing_mock.c:135]: if you're aligning these parameters now, you could just put the first parameter one line above and move all other parameters onto the second line. Same below

Comment (perahgren) [reply src/audio/google_rtc_audio_processing_mock.c:135]: True. 
Done.

File: src/audio/google_rtc_audio_processing_mock.c (hunk 20 above)
Comment (cujomalainey) [reply src/audio/google_rtc_audio_processing_mock.c:44]: @lyakh given this is test code im inclined to reject this change. If it was in the adapter then I would agree

File: src/audio/google_rtc_audio_processing.c [hunk 22]
```diff
@@ -275,6 +410,8 @@ static struct comp_dev *google_rtc_audio_processing_create(
 	if (cd) {
 		rfree(cd->output_buffer);
 		rfree(cd->aec_reference_buffer);
+		GoogleRtcAudioProcessingDetachMemoryBuffer();
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing.c:413]: where is `cd->state` cleaned up?

Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c:416]: Good find! It is not at all cleaned up here, and has never been.

I added a conditional cleanup.
PTAL.

**Review by cujomalainey — APPROVED** [review]

File: src/audio/google_rtc_audio_processing.c [hunk 23]
```diff
@@ -293,6 +433,8 @@ static void google_rtc_audio_processing_free(struct comp_dev *dev)
 	cd->state = NULL;
 	rfree(cd->output_buffer);
 	rfree(cd->aec_reference_buffer);
+	GoogleRtcAudioProcessingDetachMemoryBuffer();
+	rfree(cd->memory_buffer);
```
[the remark this answers is not in the document]
Comment (cujomalainey) [reply src/audio/google_rtc_audio_processing.c:437]: nvm im blind

**Review by cujomalainey — APPROVED** [review]

File: src/audio/google_rtc_audio_processing.c [hunk 24]
```diff
@@ -34,8 +34,9 @@
 
 #include <google_rtc_audio_processing.h>
 #include <google_rtc_audio_processing_platform.h>
+#include <google_rtc_audio_processing_sof_message_reader.h>
 
-#define GOOGLE_RTC_AUDIO_PROCESSING_SAMPLERATE 48000
+#define GOOGLE_RTC_AUDIO_PROCESSING_FRAMES_PER_SECOND 100
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing.c:39]: I think this name is false, based on the usage it looks like its better to make this macro called FREQENCY_TO_PERIOD_FRAMES

File: src/audio/google_rtc_audio_processing_mock.c [hunk 25]
```diff
@@ -13,6 +13,8 @@
 #include "ipc/topology.h"
 
 #define GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ 48000
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing_mock.c:15]: this is now a config

File: src/audio/google_rtc_audio_processing_mock.c [hunk 26]
```diff
@@ -13,6 +13,8 @@
...
+#define GOOGLE_RTC_AUDIO_PROCESSING_FRAMES_PER_SECOND 100
```
Comment (cujomalainey) [inline src/audio/google_rtc_audio_processing_mock.c:16]: this is redefined, should it go in a header?

File: src/audio/google_rtc_audio_processing_mock.c (hunk 25 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing_mock.c:15]: Well spotted. I now changed to using that.

File: src/audio/google_rtc_audio_processing_mock.c (hunk 26 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing_mock.c:16]: I would prefer to not put it into the header. google_rtc_audio_processing.h and instead redefine it here.

File: src/audio/google_rtc_audio_processing.c (hunk 24 above)
Comment (perahgren) [reply src/audio/google_rtc_audio_processing.c:39]: Sounds good.
Done.

**Review by cujomalainey — APPROVED** [review]

**Comment by cujomalainey** [comment]
Merging as this is a google isolated change. Note to readers, next week I will move this to our own folder with our own kconfig.
